src/offline/mod_offline_rdbms.erl

```diff
@@ -34,22 +34,91 @@
          remove_old_messages/2,
          remove_user/2]).
 
+-import(mongoose_rdbms, [prepare/4, execute_successfully/3]).
+
 -include("mongoose.hrl").
 -include("jlib.hrl").
 -include("mod_offline.hrl").
 
 -define(OFFLINE_TABLE_LOCK_THRESHOLD, 1000).
 
 init(_Host, _Opts) ->
+    prepare_queries(),
+    ok.
+
+prepare_queries() ->
+    prepare(offline_insert, offline_message,
+            [username, server, from_jid, timestamp, expire,
+             packet, permanent_fields],
+            <<"INSERT INTO offline_message "
+              "(username, server, from_jid, timestamp, expire,"
+              " packet, permanent_fields) "
+              "VALUES (?, ?, ?, ?, ?, ?, ?)">>),
+    {LimitSQL, LimitMSSQL} = rdbms_queries:get_db_specific_limits_binaries(),
+    prepare(offline_count_limit, offline_message,
+            rdbms_queries:add_limit_arg(limit, [server, username]),
+            <<"SELECT ", LimitMSSQL/binary,
+              " count(*) FROM offline_message "
+              "WHERE server=? AND username=? ", LimitSQL/binary>>),
+    prepare(offline_select, offline_message,
+            [server, username, expire],
+            <<"SELECT timestamp, from_jid, packet, permanent_fields "
+              "FROM offline_message "
+              "WHERE server = ? AND username = ? AND "
+                     "(expire IS null OR expire > ?) "
+              "ORDER BY timestamp">>),
+    prepare(offline_delete, offline_message,
+            [server, username],
+            <<"DELETE FROM offline_message "
+              "WHERE server = ? AND username = ?">>),
+    prepare(offline_delete_old, offline_message,
+            [timestamp],
+            <<"DELETE FROM offline_message WHERE timestamp < ?">>),
+    prepare(offline_delete_expired, offline_message,
+            [expire],
+            <<"DELETE FROM offline_message "
+              "WHERE expire IS NOT null AND expire < ?">>),
     ok.
 
+execute_count_offline_messages(LUser, LServer, Limit) ->
+    Args = rdbms_queries:add_limit_arg(Limit, [LServer, LUser]),
+    execute_successfully(LServer, offline_count_limit, Args).
+
+execute_fetch_offline_messages(LServer, LUser, ExtTimeStamp) ->
+    execute_successfully(LServer, offline_select, [LServer, LUser, ExtTimeStamp]).
+
+execute_remove_expired_offline_messages(LServer, ExtTimeStamp) ->
+    execute_successfully(LServer, offline_delete_expired, [ExtTimeStamp]).
+
+execute_remove_old_offline_messages(LServer, ExtTimeStamp) ->
+    execute_successfully(LServer, offline_delete_old, [ExtTimeStamp]).
+
+execute_remove_user(LServer, LUser) ->
```

Comment:
Could it be renamed? It does not remove the user, but their offline messages.

Reply:
We have `remove_user` hook callback, which does not remove the user, but their offline messages...